chainer/functions/activation/relu.py

```diff
@@ -24,36 +25,94 @@ def check_type_forward(self, in_types):
         )
 
     def forward_cpu(self, x):
-        self.retain_inputs(())
         self.retain_outputs((0,))
         return utils.force_array(numpy.maximum(x[0], 0, dtype=x[0].dtype)),
 
     def forward_gpu(self, x):
         if chainer.should_use_cudnn('==always') and x[0].flags.c_contiguous:
+            # cupy.activation_backward requires the input.
+            # So, we retain it for backward computation.
+            self.retain_inputs((0,))
             self._use_cudnn = True
             y = cudnn.activation_forward(x[0], _mode)
         else:
-            self.retain_inputs(())
-            self._use_cudnn = False
             y = cuda.cupy.maximum(x[0], 0)
         self.retain_outputs((0,))
         return y,
 
-    def backward_cpu(self, x, gy):
-        y = self.output_data[0]
-        return utils.force_array(gy[0] * (y > 0)),
-
-    def backward_gpu(self, x, gy):
-        y = self.output_data[0]
+    def backward(self, indexes, gy):
+        y = self.get_retained_outputs()[0]
         if chainer.should_use_cudnn('==always') and self._use_cudnn:
-            gx = cudnn.activation_backward(x[0], y, gy[0], _mode)
+            x = self.get_retained_inputs()[0]
+            return ReLUGrad3(x, y).apply((gy[0],))
         else:
-            gx = cuda.elementwise(
-                'T y, T gy', 'T gx',
-                'gx = y > 0 ? gy : (T)0',
-                'relu_bwd')(y, gy[0])
+            return ReLUGrad2(y).apply((gy[0],))
+
+
+def _heaviside(x):
+    return (x > 0).astype(x.dtype)
+
+
+class ReLUGrad2(function_node.FunctionNode):
+    """Computes the gradient of the ReLU function.
+
+    This function takes 2 variables b and c, and
+    computes f(b, c) = sign(b) * c with backpropagation
+    where operations are dones in elementwise manner
+    and sign(x) = 1 when x > 0 is positive and 0 otherwise.
+
+    As the gradient of f with respect to b is 0,
+    we do not backpropagate errors toward b for computational efficiency.
+    """
+
+    def __init__(self, b):
+        super(ReLUGrad2).__init__()
```

Comment:
`, self` is missing (or remove this line; it is allowed to not call super init in FunctionNode)